---
title: "00.00-fastqc-multiqc-raw-reads"
author: "Sam White"
date: "2025-04-17"
output: 
  github_document:
    toc: true
    number_sections: true
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
---

# BACKGROUND

This notebook will download raw sequencing reads from https://owl.fish.washington.edu/nightingales/eDNA-yellow_island-2023/ and run an initial quality assessement using [FastQC](https://github.com/s-andrews/FastQC) and [MultiQC](https://github.com/MultiQC/MultiQC) [@ewels2016].

```{r setup, include=FALSE}
library(knitr)
library(dplyr)
library(reticulate)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

Inputs:

- Paired-end FastQs

  - Example filename format: `Yell_July18_Site1f_S92_L001_R[12]_001.fastq.gz`

Outputs:

- FastQC HTML files.

- MultiQC HTML report.

Software requirements:

-   [FastQC](https://github.com/s-andrews/FastQC)

-   [MultiQC](https://github.com/MultiQC/MultiQC) [@ewels2016]

# Set R variables

```{r R-variables, eval=TRUE}

# Data directories
data_dir <- "../data/"
raw_reads_dir <- "../data/raw-fastqs"
raw_reads_url <- "https://owl.fish.washington.edu/nightingales/eDNA-yellow_island-2023/"

# PROGRAMS
fastqc <- "/home/sam/programs/FastQC-0.12.1/fastqc"
multiqc <- "/home/sam/programs/miniforge3/bin/multiqc"

# CPUs
threads <- 4


# Export these as environment variables for bash chunks.
Sys.setenv(
  fastqc = fastqc,
  data_dir = data_dir,
  multiqc = multiqc,
  raw_reads_dir = raw_reads_dir,
  raw_reads_url = raw_reads_url,
  threads = threads
)
```

# Download raw FastQs

The `--cut-dirs 3` command cuts the preceding directory structure (i.e. `nightingales/eDNA-yellow_island-2023/`)
so that we just end up with the reads.

```{r download-raw-fastqs, engine='bash', eval=TRUE}
# Make output directory if it doesn't exist
mkdir --parents "${raw_reads_dir}"

wget \
--directory-prefix "${raw_reads_dir}" \
--recursive \
--no-check-certificate \
--continue \
--cut-dirs 3 \
--no-host-directories \
--no-parent \
--quiet \
--accept "*.fastq.gz,*.md5" \
"${raw_reads_url}"

ls -lh "${raw_reads_dir}"
```

## Verify raw read checksums

Removes MD5 files after check to facilitate downstream directory parsing by QIIME 2.
```{r verify-raw-read-checksums, engine='bash', eval=TRUE}

cd "${raw_reads_dir}"

for md5 in *.md5
do
  md5sum --check "${md5}"
  rm "${md5}"
done
```



# FastQC/MultiQC on raw reads

```{r raw-fastqc-multiqc, engine='bash', eval=TRUE}

# Make output directory if it doesn't exist
mkdir --parents "${raw_reads_dir}"


############ RUN FASTQC ############


# Create array of trimmed FastQs
# Access the individual FastQ files from the array
IFS=' ' read -r -a raw_fastqs_array <<< "${raw_fastqs_string}"

# Pass array contents to new variable as space-delimited list
raw_fastqc_list=$(echo "${raw_fastqs_array[*]}")


echo "Beginning FastQC on raw reads..."
echo ""

# Run FastQC
### NOTE: Do NOT quote raw_fastqc_list
${fastqc} \
--threads ${threads} \
--outdir ${raw_reads_dir} \
--quiet \
${raw_fastqc_list}

echo "FastQC on raw reads complete!"
echo ""

############ END FASTQC ############

############ RUN MULTIQC ############
echo "Beginning MultiQC on raw FastQC..."
echo ""

${multiqc} --interactive \
${raw_reads_dir} \
-o ${raw_reads_dir}

echo ""
echo "MultiQC on raw FastQs complete."
echo ""

############ END MULTIQC ############

echo "Removing FastQC zip files."
echo ""
rm ${raw_reads_dir}/*.zip
echo "FastQC zip files removed."
echo ""
```
# Check for empty/mismatched FastQs

## Delete empty FastQs
```{r delete-zero-reads, eval=TRUE}
# Path to the fastqc_sequence_counts_plot.txt file
counts_file <- file.path(Sys.getenv("raw_reads_dir"), "multiqc_data", "fastqc_sequence_counts_plot.txt")

# Check if the file exists
if (file.exists(counts_file)) {
  # Read the file
  counts_data <- read.table(counts_file, header = TRUE, sep = "\t", stringsAsFactors = FALSE)

  # Identify samples with Unique Reads equal to 0.0
  zero_reads_samples <- counts_data$Sample[counts_data$Unique.Reads == 0.0]

  # Check if any samples with zero reads were found
  if (length(zero_reads_samples) > 0) {
    cat("FastQ files with zero unique reads found:\n")

    # Loop through the samples and delete the corresponding FastQ files
    for (sample in zero_reads_samples) {
      # Construct the file path
      fastq_file <- file.path(Sys.getenv("raw_reads_dir"), paste0(sample, ".fastq.gz"))

      # Check if the file exists
      if (file.exists(fastq_file)) {
        # Delete the file
        unlink(fastq_file)
        cat("Deleted:", fastq_file, "\n")
      } else {
        cat("File not found:", fastq_file, "\n")
      }
    }
  } else {
    cat("No FastQ files with zero unique reads found.\n")
  }
} else {
  cat("Error: fastqc_sequence_counts_plot.txt file not found at", counts_file, "\n")
}
```

## Check for missing FastQ pairs
```{r missing-fastq-pairs, eval=TRUE}
# Create the fastq list and string
raw_fastqs_list <- list.files(path = Sys.getenv("raw_reads_dir"), pattern = "\\.fastq\\.gz$", full.names = TRUE)
raw_fastqs_string <- paste(raw_fastqs_list, collapse = " ")

# Update the environment variable with the actual string
Sys.setenv(raw_fastqs_string = raw_fastqs_string)

# Check for matching R1 and R2 files
r1_files <- grep("_R1_", raw_fastqs_list, value = TRUE)
r2_files <- grep("_R2_", raw_fastqs_list, value = TRUE)

# Extract sample names from R1 and R2 files
r1_samples <- gsub("_R1_.*", "", basename(r1_files))
r2_samples <- gsub("_R2_.*", "", basename(r2_files))

# Find unmatched R1 files
unmatched_r1 <- setdiff(r1_samples, r2_samples)
if (length(unmatched_r1) > 0) {
  cat("Unmatched R1 files:\n")
  for (sample in unmatched_r1) {
    r1_file <- grep(paste0("^", sample, "_R1_"), basename(r1_files), value = TRUE)
    r1_file_path <- file.path(Sys.getenv("raw_reads_dir"), r1_file)
    cat(r1_file, "\n")
    # Delete unmatched R1 file
    unlink(r1_file_path)
    cat("Deleted:", r1_file_path, "\n")
  }
}

# Find unmatched R2 files
unmatched_r2 <- setdiff(r2_samples, r1_samples)
if (length(unmatched_r2) > 0) {
  cat("Unmatched R2 files:\n")
  for (sample in unmatched_r2) {
    r2_file <- grep(paste0("^", sample, "_R2_"), basename(r2_files), value = TRUE)
    r2_file_path <- file.path(Sys.getenv("raw_reads_dir"), r2_file)
    cat(r2_file, "\n")
    # Delete unmatched R2 file
    unlink(r2_file_path)
    cat("Deleted:", r2_file_path, "\n")
  }
}

```

```{r check-file-listing, engine='bash', eval=TRUE}

# View directory contents
ls -lh ${raw_reads_dir}

```
