---
title: "02.00-qiime2"
author: "Sam White"
date: "2025-04-16"
output: 
  github_document:
    toc: true
    number_sections: true
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
---

# BACKGROUND

This notebook will use [QIIME 2 2024.10](https://docs.qiime2.org/2024.10/) [@bolyen2019] to taxonomically classify eukaryotic sequences in eDNA samples.

```{r setup, include=FALSE}
library(knitr)
library(dplyr)
library(reticulate)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

Inputs:

- Trimmed FastQs from [`01.00-trimming-fastp-fastqc-multiqc.Rmd`](https://github.com/RobertsLab/project-eDNA-yellow/blob/9d2b4a6a77820edb0c3b7a78878a8c7ef6a61f61/code/01.00-trimming-fastp-fastqc-multiqc.Rmd)

  - Filename format: `Yell_July18_Site1L_S101_L001_R1.fastp-trim.fastq.gz`


Outputs:



Software requirements:

-   Utilizes a [`qiime2-amplicon-2024.10`](https://docs.qiime2.org/2024.10/) Conda/Mamba environment, per the installation instructions.

To use on your own system, replace with name of your qiime2-amplicon-2024.10 environment and the path to the corresponding conda installation (find this *after* you've activated the environment).

E.g.

``` bash
# Activate environment
conda activate ShortStack4_env

# Find conda path
which conda
```

------------------------------------------------------------------------

# Set R variables

```{r R-variables, eval=TRUE}
# Conda info
qiime2_conda_env_name <- c("qiime2-amplicon-2024.10")
qiime2_conda_path <- c("/home/sam/programs/miniforge3/condabin/conda")

# Data directories
repo_dir <- "/home/sam/gitrepos/RobertsLab/project-eDNA-yellow"
data_dir <- file.path(repo_dir, "data")
trimmed_reads_dir <- file.path(repo_dir, "output", "01.00-trimming-fastp-fastqc-multiqc")

# Output files/directories
output_dir <- file.path(repo_dir, "output", "02.00-qiime2")

# Export these as environment variables for bash chunks.
Sys.setenv(
  data_dir = data_dir,
  trimmed_reads_dir = trimmed_reads_dir,
  output_dir = output_dir
)

```

# Load [`qiime2-amplicon-2024.10`](https://docs.qiime2.org/2024.10/) conda environment

If this is successful, the first line of output should show that the Python being used is the one in your [`qiime2-amplicon-2024.10`](https://docs.qiime2.org/2024.10/) conda environment path.

E.g.

`python:         /home/sam/programs/mambaforge/envs/mirmachine_env/bin/python`

```{r load-qiim2-conda-env, eval=TRUE}
use_condaenv(condaenv = qiime2_conda_env_name, conda = qiime2_conda_path)

# Check successful env loading
py_config()
```



# Import trimmed reads
```{r import-trimmed-reads, engine=bash, eval=TRUE}
# Make output directory if it doesn't exist
mkdir --parents "${output_dir}"

qiime tools import \
  --type 'SampleData[PairedEndSequencesWithQuality]' \
  --input-path "${trimmed_reads_dir}" \
  --input-format CasavaOneEightSingleLanePerSampleDirFmt \
  --output-path "${output_dir}"/demux-paired-end.qza

```

# Deblur
```{r deblur, engine=bash, eval=TRUE}
qiime quality-filter q-score \
 --i-demux "${output_dir}"/demux-paired-end.qza \
 --o-filtered-sequences "${output_dir}"/demux-paired-end-filtered.qza \
 --o-filter-stats "${output_dir}"/demux-paired-end-filter-stats.qza
```