---
title: "02.00-qiime2"
author: "Sam White"
date: "2025-04-16"
output: 
  github_document:
    toc: true
    number_sections: true
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
---

# BACKGROUND

This notebook will use [QIIME 2 2024.10](https://docs.qiime2.org/2024.10/) [@bolyen2019] to taxonomically classify eukaryotic sequences in eDNA samples. Expects trimmed reads to exist in [`../output/01.00-trimming-fastp-fastqc-multiqc`](https://github.com/RobertsLab/project-eDNA-yellow/tree/main/output/01.00-trimming-fastp-fastqc-multiqc).

```{r setup, include=FALSE}
library(knitr)
library(dplyr)
library(reticulate)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

## Inputs:

- Trimmed FastQs from [`01.00-trimming-fastp-fastqc-multiqc.Rmd`](https://github.com/RobertsLab/project-eDNA-yellow/blob/13bdbbec60b0da23266af2e0b409c2b09f3d38fa/code/01.00-trimming-fastp-fastqc-multiqc.Rmd)

  - Filename format: `fastp-Yell-July18-Site5e_S95_L001_R1_001.fastq.gz`
  
- SILVA classifier QZA `silva-138-99-nb-classifier.qza`.

  - Downloaded from https://library.qiime2.org/data-resources#naive-bayes-classifiers


## Outputs:

Files ending in `.qza` are  QIIME zip archives.

Files ending in `.qzv` are QIIME zip visualization tables.

- `deblur-stats.qza`: Stats after denoise.

- `deblur-stats.qzv`: Stats visualization after denoise.

- `demux-paired-end-filtered.qza`: Quality score filtered archive.

- `demux-paired-end-filter-stats.qza`: Stats archive for quality score filtering.

- `demux-paired-end.qza`: Paired-end reads archive.

- `rep-seqs-deblur.qza`: Representative sequences after denoise.

- `rep-seqs-deblur.qzv`: Representative sequences after denoise visualization.

- `table-deblur.qza`: Table after denoise.

- `table.qzv`: Feature table summary visuallization.

- `taxa-bar-plots.qzv`: Bar plots of taxonomic assignment visualization.

- `taxonomy.qza`: Feature taxonomic classification after applying SILVA classifier.

- `taxonomy.qzv`: Feature taxonomic classification visualization after applying SILVA classifier.


Software requirements:

-   Utilizes a [`qiime2-amplicon-2024.10`](https://docs.qiime2.org/2024.10/) Conda/Mamba environment, per the installation instructions.

To use on your own system, replace with name of your qiime2-amplicon-2024.10 environment and the path to the corresponding conda installation (find this *after* you've activated the environment).

E.g.

``` bash
# Activate environment
conda activate ShortStack4_env

# Find conda path
which conda
```

------------------------------------------------------------------------

# Set R variables

Change `repo_dir` to match local repo location.

Change `threads <- 4` to match computer resources for your system.

```{r R-variables, eval=TRUE}
# Conda info
qiime2_conda_env_name <- c("qiime2-amplicon-2024.10")
qiime2_conda_path <- c("/home/sam/programs/miniforge3/condabin/conda")

# Data directories
repo_dir <- "/home/sam/gitrepos/RobertsLab/project-eDNA-yellow"
data_dir <- file.path(repo_dir, "data")
output_dir <- file.path(repo_dir, "output", "02.00-qiime2")
trimmed_reads_dir <- file.path(repo_dir, "output", "01.00-trimming-fastp-fastqc-multiqc")

# Files
metadata <- file.path(data_dir, "raw-fastqs", "metadata.tsv")
silva_classifier <- file.path(data_dir, "silva-138-99-nb-classifier.qza")
silva_ref_seqs <- file.path(data_dir, "silva-138-99-seqs.qza")

# Parameters

## Set Deblur read length
## Based on FastQC results, post-trimming, where quality starts to decline.
deblur_ptrim_length <- 150

# CPUs
threads <- 4


# Export these as environment variables for bash chunks.
Sys.setenv(
  data_dir = data_dir,
  deblur_ptrim_length = deblur_ptrim_length,
  metadata = metadata,
  silva_classifier = silva_classifier,
  silva_ref_seqs = silva_ref_seqs,
  threads = threads,
  trimmed_reads_dir = trimmed_reads_dir,
  output_dir = output_dir
)

```

# Load [`qiime2-amplicon-2024.10`](https://docs.qiime2.org/2024.10/) conda environment

If this is successful, the first line of output should show that the Python being used is the one in your [`qiime2-amplicon-2024.10`](https://docs.qiime2.org/2024.10/) conda environment path.

E.g.

`python:         /home/sam/programs/mambaforge/envs/mirmachine_env/bin/python`

```{r load-qiim2-conda-env, eval=TRUE}
use_condaenv(condaenv = qiime2_conda_env_name, conda = qiime2_conda_path)

# Check successful env loading
py_config()
```


# Import trimmed reads

## Transfer trimmed reads
```{r transfer-trimmed-reads, engine='bash', eval=TRUE}
# Make output directory if it doesn't exist
mkdir --parents "${output_dir}"

rsync -avP "${trimmed_reads_dir}"/*.fastq.gz "${output_dir}"

ls -lh "${output_dir}"
```

## Import to QIIME2 QZA format
```{r import-trimmed-reads, engine='bash', eval=TRUE}

qiime tools import \
  --type 'SampleData[PairedEndSequencesWithQuality]' \
  --input-path "${output_dir}" \
  --input-format CasavaOneEightSingleLanePerSampleDirFmt \
  --output-path "${output_dir}"/demux-paired-end.qza
  
# Remove trimmed reads
rm "${output_dir}"/*.fastq.gz

```

# Deblur

## Quality filter
```{r deblur, engine='bash', eval=TRUE}
qiime quality-filter q-score \
 --i-demux "${output_dir}"/demux-paired-end.qza \
 --o-filtered-sequences "${output_dir}"/demux-paired-end-filtered.qza \
 --o-filter-stats "${output_dir}"/demux-paired-end-filter-stats.qza
```

## Denoise

Trim length is based on FastQC results,
where most samples show sharp drop in quality scores.

```{r deblur-denoise, engine='bash', eval=TRUE}
qiime deblur denoise-other \
  --i-demultiplexed-seqs "${output_dir}"/demux-paired-end-filtered.qza \
  --i-reference-seqs "${silva_ref_seqs}" \
  --p-trim-length "${deblur_ptrim_length}" \
  --o-representative-sequences "${output_dir}"/rep-seqs-deblur.qza \
  --o-table "${output_dir}"/table-deblur.qza \
  --p-sample-stats \
  --p-jobs-to-start "${threads}" \
  --o-stats "${output_dir}"/deblur-stats.qza
```

## Deblur visualization
```{r deblur-visualization, engine='bash', eval=TRUE}
qiime deblur visualize-stats \
  --i-deblur-stats "${output_dir}"/deblur-stats.qza \
  --o-visualization "${output_dir}"/deblur-stats.qzv
```

# Feature Table
```{r feature-table, engine='bash', eval=TRUE}
qiime feature-table summarize \
  --i-table "${output_dir}"/table-deblur.qza \
  --o-visualization "${output_dir}"/table.qzv \
  --m-sample-metadata-file "${metadata}"

qiime feature-table tabulate-seqs \
  --i-data "${output_dir}"/rep-seqs-deblur.qza \
  --o-visualization "${output_dir}"/rep-seqs-deblur.qzv

```

# Taxonomy

## Classification

```{r taxonomic-classification, engine='bash', eval=TRUE}
qiime feature-classifier classify-sklearn \
  --i-classifier "${silva_classifier}" \
  --i-reads "${output_dir}"/rep-seqs-deblur.qza \
  --o-classification "${output_dir}"/taxonomy.qza

qiime metadata tabulate \
  --m-input-file "${output_dir}"/taxonomy.qza \
  --o-visualization "${output_dir}"/taxonomy.qzv
```

## Bar plots
```{r taxonomic-bar-plots, engine='bash', eval=TRUE}
qiime taxa barplot \
  --i-table "${output_dir}"/table-deblur.qza \
  --i-taxonomy "${output_dir}"/taxonomy.qza \
  --m-metadata-file "${metadata}" \
  --o-visualization "${output_dir}"/taxa-bar-plots.qzv


```

